### ISSUE TITLE
Support Kernels for SVM

### ISSUE THREAD

**Issue by Rikorose** [tensorflow/tensorflow#8904]
Kernels in support vector machines are really usefull. I would like to test some object detection algorithms in tensorflow using svm and a radial basis kernel.

I am willing to help implementing kernels for support vector machines, though I am not familiar with the sdca optimization. 

Is there any design doc, or plan how to add more features to the svm estimator api?

**Comment by aselle** [comment]
You might if anybody has done anything in this vein already by asking on stackoverflow. I don't know of any work internally to build out the svm functionality more or any design doc.
@martinwicke, do you have any knowledge of others interested in this or a good course of action? Marking as contributions welcome for now.

**Comment by martinwicke** [comment]
@petrosmol should have the best context for this. I believe there is work in progress for kernel-SVMs.

**Comment by petrosmol** [comment]
TL;DR: Yes, we are actively working on kernel methods in general. As of last week there is a [kernel_methods](https://github.com/tensorflow/tensorflow/tree/master/tensorflow/contrib/kernel_methods) package which is part of tensorflow.contrib. Kernel SVMs are supported indirectly (with some more work on user's behalf, see below). Note that this is all very recent so you need to build from source (and some changes might still need some time to propagate). Read on for details.

* We don't have support for *exact* kernel methods and AFAIK, there is no plan to support these any time soon. However, we do support primal (approximate) kernel methods. We have implemented Random Fourier Features (RFF) that approximate the RBF kernel. This is the equivalent of sklearn's RBFSampler. See [here](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/contrib/kernel_methods/python/mappers/random_fourier_features.py). You can use RFF to transform your input features and then apply a linear model on top of it. These are typically much faster than exact kernels and perform pretty well in practice (quality-wise). Note that this works with dense input features only (so images should be fine)
* If you want to use a pre-packaged estimator, we have implemented a Linear Classifier that uses RFF to preprocess the input features. This is not kernel SVMs, you can think of it as a "kernelization" of the LinearClassifier model from tensorflow.contrib.learn. The loss used is logistic (for binary) and cross entropy (for multiclass). See [here](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/contrib/kernel_methods/python/kernel_estimators.py) for the estimator and [here](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/contrib/kernel_methods/python/kernel_estimators_test.py) for sample usage.
* If you really want to SVMs, you have the following 2 options:
   * If your problem is binary classification you can use the pre-packed SVM estimator (which uses SDCA) but you should first apply [RFF](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/contrib/kernel_methods/python/mappers/random_fourier_features.py) to your features (as part of your input_fn). Make sure you update your feature columns accordingly.
   * For multiclass, you can implement your own 1-vs-all or 1-vs-1 using binary SVMs. There is no multiclass support for SVMs currently (but there are plans for it)

**Comment by Rikorose** [comment]
Thank you! I will try the different options. Binary SVM is sufficient for me right now.

**Comment by Lidaguo** [comment]
Hello, @petrosmol !  Has multiclass SVMs api been supported in TF version 1.6 now?

**Comment by petrosmol** [comment]
There is no pre-packaged multiclass SVM Estimator yet (it is in the future plans though). But since this bug was opened, I have implemented multiclass SVM loss (see https://github.com/tensorflow/tensorflow/blob/master/tensorflow/contrib/kernel_methods/python/losses.py). If you want to turn this into an Estimator, possibly the easiest way is to create a class that inherits from [_KernelEstimator](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/contrib/kernel_methods/python/kernel_estimators.py) and pass to its constructor a head that uses the multiclass SVM loss (see KernelLinearClassifier model for instance that uses a cross-entropy multiclass loss)

**Comment by Lidaguo** [comment]
Thank you very much! @petrosmol  I will try your advice.  :-  )

**Comment by Lidaguo** [comment]
Hello, @petrosmol ! I would like to know whether the implemented multiclass SVM loss  (see [https://github.com/tensorflow/tensorflow/blob/master/tensorflow/contrib/kernel_methods/python/losses.py]) in tensorflow could be optimized by Gradient Descent Algorithm.

**Comment by petrosmol** [comment]
@Lidaguo. Yes, the loss should work with any gradient based optimizer. I have tried it in the past in some experiments with MNIST with no issues AFAIR.

**Comment by Lidaguo** [comment]
Hello, @petrosmol ! Thanks for your particular reply. But now, I have another problem. In my case, I'd like to use the implemented tf.contrib.kernel_methods.RandomFourierFeatureMapper to transform the features from network, and then use the transformed features to construct my loss function.  All the things go well at the beginning of my training procedure. However,  with the increasing of the training epochs, the occupation of GPU grow up. After tens of epochs, the OOM error of GPU occurs. If I don't use the transformation, everything works fine. my simple code as following:
`
kernel_mapper = tf.contrib.kernel_methods.RandomFourierFeatureMapper(input_dim = 100, output_dim = 1000,  stddev =5.0, name='rffm')
model_train = My_model(..., kernel_mapper)
for epoch_index in xrange(total_epochs):
    run_epoch(..., model_train)
`
Here is the definition of the model,
`
class My_model(..., kernel_mapper):
     ....
     input_feature = ....
     mapped_feature = kernel_mapper.map(input_feature)
    ....
`
I guess that the  tf.contrib.kernel_methods.RandomFourierFeatureMapper class always stores values for each epoch(or each called situation) and don't free the Memory. Is there any solution for this?

**Comment by Lidaguo** [comment]
Hello, @petrosmol ! Thanks for your particular reply. But now, I have another problem. In my case, I'd like to use the implemented tf.contrib.kernel_methods.RandomFourierFeatureMapper to transform the features from network, and then use the transformed features to construct my loss function.  All the things go well at the beginning of my training procedure. However,  with the increasing of the training epochs, the occupation of GPU grow up. After tens of epochs, the OOM error of GPU occurs. If I don't use the transformation, everything works fine. my simple code as following:
`
kernel_mapper = tf.contrib.kernel_methods.RandomFourierFeatureMapper(input_dim = 100, output_dim = 1000,  stddev =5.0, name='rffm')

model_train = My_model(..., kernel_mapper)

for epoch_index in xrange(total_epochs):
    run_epoch(..., model_train)

`
Here is the definition of the model,

`
class My_model(..., kernel_mapper):
     ....
     input_feature = ....
     mapped_feature = kernel_mapper.map(input_feature)
    ....

`
I guess that the  tf.contrib.kernel_methods.RandomFourierFeatureMapper class always stores values for each epoch(or each called situation) and don't free the Memory. Is there any solution for this?

**Comment by petrosmol** [comment]
@Lidaguo.

Hi, RandomFourierFeatureMapper should be able to handle input/output dimensions of 100/1000. Related question: Do you see the same issue on CPU?

Is it possible that you add the map ops multiple times (one per epoch) in your training graph and that's why your are getting OOM errors after a few iterations? 
RandomFourierFeatureMapper only adds map OPs (and does not store any values in a way that is different than what any TF graph would do it).

Can you make sure that, under the hood, your call does not call .map method multiple times?

For debugging, there are ways to print the size of the TF graph after each iteration/epoch. If this is growing with each iteration then you are likely adding ops to your graph during training (which is bad).

If the above suggestions don't help you fix the problem, send me the full code so that I can reproduce the issue.

Thanks
Petros

**Comment by Lidaguo** [comment]
@petrosmol . Hello, I did add the map ops multiple times in one per epoch in my training graph. I have fixed this problem. Thank you very much. And I have a another problem about the implemented multiclass SVM loss (see [)](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/contrib/kernel_methods/python/losses.py)

**Comment by Lidaguo** [comment]
@petrosmol . Hello, I did add the map ops multiple times in one per epoch in my training graph. I have fixed this problem. Thank you very much. And now, I have a problem about the implemented multiclass SVM loss (see [https://github.com/tensorflow/tensorflow/blob/master/tensorflow/contrib/kernel_methods/python/losses.py](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/contrib/kernel_methods/python/losses.py)), like the following picture:
![e](https://user-images.githubusercontent.com/15832113/38450361-03dc76d6-3a4f-11e8-8010-780365f9e72f.png)

In my case, 
`self._labels = tf.placeholder(tf.int32, [None,],name='data_labels')`
Dose it mean that the **_labels_** or **_logits_** sent into the multiclass SVM loss class  must be a constant tensor ?

**Comment by petrosmol** [comment]
@Lidaguo 
Hard to tell what is going on without seeing code. But both labels and logits should work with unspecified shape. For instance the following code should work:

import numpy as np

BATCH_SIZE = None
NUM_CLASSES = None

logits_np = np.array([[1.2, -1.4, -1.0], [1.4, 1.8, 4.0], [0.5, 1.3, -1.0]])
labels_np = np.array([0, 2, 1], dtype=np.int32)
logits = tf.placeholder(
    tf.float32, shape=(BATCH_SIZE, NUM_CLASSES), name='logits')
labels = tf.placeholder(tf.int32, shape=[BATCH_SIZE,], name='labels')
loss = tf.contrib.kernel_methods.sparse_multiclass_hinge_loss(labels, logits)

with tf.Session() as sess:
  loss_eval = sess.run(loss, feed_dict={logits: logits_np, labels: labels_np})
  print(loss_eval)

Initial version of kernel_methods/python/losses.py was not working with unknown shape though. This was fixed end of 2017 (see https://github.com/tensorflow/tensorflow/commit/2dce1ee2e37e314772467bfa150b57b8d4311598#diff-c7afce59582d7278b51608c8024bf615)

So, you might need to upgrade to a more recent version of tf or patch these changes locally.

**Comment by Lidaguo** [comment]
@petrosmol 
Yes! I have ungraded the version of tf. Now, my model works. Thank you very much. : - )
